src/lib/rules/jsll-script-included/jsll-script-included.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * @fileoverview This rule confirms that JSLL script is included in the page
+ */
+import { Category } from '../../enums/category';
+import { RuleContext } from '../../rule-context';
+import { IAsyncHTMLDocument, IAsyncHTMLElement, IRule, IRuleBuilder, ITraverseEnd, Severity } from '../../types';
+import { isHTMLDocument } from '../../utils/misc';
+
+/*
+ * ------------------------------------------------------------------------------
+ * Public
+ * ------------------------------------------------------------------------------
+ */
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        const validate = async (data: ITraverseEnd) => {
+            const { resource }: { resource: string } = data;
+            const jsllDir = `https://az725175.vo.msecnd.net/scripts/jsll-`;
+            const passRegex = new RegExp(`^(\\d+\\.)js`);
+            const warningRegex = new RegExp(`^(\\d+\\.){2,}js`);
+
+            // Messages.
+            const noScriptInHeadMsg = `No JSLL script was included in the <head> tag.`;
+            const redundantScriptInHeadMsg = `More than one JSLL scripts were included in the <head> tag.`;
+            const warningScriptVersionMsg = `Use the latest release of JSLL with 'jsll-4.js'. It is not recommended to specify the version number unless you wish to lock to a specific release.`;
+            const invalidScriptVersionMsg = `The jsll script versioning is not valid.`;
+            const wrongScriptOrderMsg = `The JSLL script isn't placed prior to other scripts.`;
+
+            const getJsllScripts= (elements: Array<IAsyncHTMLElement>) => {
+                return elements.filter((element) => {
+                    return element.getAttribute('src').startsWith(jsllDir);
+                });
+            };
+
+            if (!isHTMLDocument(resource, context.pageHeaders)) {
+                return;
+            }
+
+            const pageDOM: IAsyncHTMLDocument = context.pageDOM as IAsyncHTMLDocument;
+            const scriptsInHead: Array<IAsyncHTMLElement> = await pageDOM.querySelectorAll('head script');
```

Comment:
We should be using the `element::script` event instead of `traverse::end`.
This should make things slightly faster in some connectors because you will not have to query the DOM to get the elements and you make a better of use of our API.

Reply:
@molant Hmm... How do you verify that the script is in `head` then?

Reply:
Vía the traverse::up/down events. If you need more information in those events to make it easier let me know.

Reply:
@molant I split the `validate` function into two validate functions, and attach them to `element::head` and `element::script` separately. By listening to the `element::head` event,  it checks the number/location of the jsll scripts without throwing duplicate messages. Does this work?

Reply:
You are still doing `querySelectorAll` at the page level. If you prefer to continue down this path then do it from the element (do we support `querySelectorAll` at the element level?). If not it is just the same thing as before but being executed in another event.

What I'd do is something like:

1. subscribe to head event or `traverse::down` to the heard
1. set a property `inHead = true`
1. in the `element::script` event, check if you are in the head via the `inHead` property. If not, quit
1. do all the validations you need to do there
1. when you do the `traverse::up`, set `inHead` to `false`

something like:

```js
let inHead = false;

const traverse = (data) =>{
  if( traverseDown of head) {
    inHead = true;
    return;
  }
 
  if (traverseUp of head) {
    inHead = false;
    return;
  }
};

const validateScript = (data) => {
  // validations here taking into account the value of inHead
};
```

Does it make sense?